numpy/core/src/multiarray/methods.c

```diff
@@ -617,23 +618,29 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
 
     fd = npy_PyFile_Dup2(file, "wb", &orig_pos);
     if (fd == NULL) {
-        goto fail;
-    }
-    if (PyArray_ToFile(self, fd, sep, format) < 0) {
-        goto fail;
+        Py_DECREF(file);
+        return NULL;
     }
-    if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
-        goto fail;
+    res_write = PyArray_ToFile(self, fd, sep, format);
+    PyErr_Fetch(&err_type, &err_value, &err_traceback);
+
+    res_close = npy_PyFile_DupClose2(file, fd, orig_pos);
+    if (res_close < 0) {
+        npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
```

Comment:
```suggestion
    int res_write = PyArray_ToFile(self, fd, sep, format);
    PyObject *err_type, *err_value, *err_traceback;
    PyErr_Fetch(&err_type, &err_value, &err_traceback);

    int res_close = npy_PyFile_DupClose2(file, fd, orig_pos);
    if (res_close < 0) {
        npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
```
There's also no need to initialize the `err_*` variables, `PyErr_Fetch` is guaranteed to write to them.